pkg/groupmapper/groupmapper.go

```diff
@@ -0,0 +1,207 @@
+	groupsCache         *usercache.GroupCache
+	groupsSynced        func() bool
+}
+
+func NewUserGroupsMapper(delegate authapi.UserIdentityMapper, groupsClient userclient.GroupInterface, groupInformer userinformer.GroupInformer) *UserGroupsMapper {
+	return &UserGroupsMapper{
+		delegatedUserMapper: delegate,
+		groupsClient:        groupsClient,
+		groupsCache:         usercache.NewGroupCache(groupInformer),
+		groupsSynced:        groupInformer.Informer().HasSynced,
+	}
+}
+
+func (m *UserGroupsMapper) UserFor(identityInfo authapi.UserIdentityInfo) (kuser.Info, error) {
+	userInfo, err := m.delegatedUserMapper.UserFor(identityInfo)
+	if err != nil {
+		return userInfo, err
+	}
+
+	identityGroups := sets.NewString(identityInfo.GetProviderGroups()...)
+	if err := m.processGroups(identityInfo.GetProviderName(), identityInfo.GetProviderPreferredUserName(), identityGroups); err != nil {
+		return nil, err
+	}
+
+	return &UserInfoGroupsWrapper{
+		userInfo:         userInfo,
+		additionalGroups: identityGroups,
+	}, nil
+}
+
+func (m *UserGroupsMapper) processGroups(idpName, username string, groups sets.String) error {
+	err := wait.PollImmediate(1*time.Second, 5*time.Second, func() (bool, error) {
+		return m.groupsSynced(), nil
+	})
+	if err != nil {
+		return err
+	}
+
+	cachedGroups, err := m.groupsCache.GroupsFor(username)
+	if err != nil {
+		return err
+	}
+
+	removeGroups, addGroups := groupsDiff(cachedGroups, groups)
+	for _, g := range removeGroups {
+		if err := m.removeUserFromGroupWithRetries(idpName, username, g); err != nil {
+			return err
+		}
+	}
+
+	for _, g := range addGroups {
+		if err := m.addUserToGroupWithRetries(idpName, username, g); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// FIXME: work better with annotations
+// FIXME: add retries
+// FIXME: use idpName
+func (m *UserGroupsMapper) removeUserFromGroupWithRetries(idpName, username, group string) error {
+	updatedGroup, err := m.groupsClient.Get(context.TODO(), group, metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+
+	if len(updatedGroup.Users) == 0 {
+		return nil
+	}
+
+	if len(updatedGroup.Users) == 1 && updatedGroup.Users[0] == username && updatedGroup.Annotations[groupGeneratedKey] == "true" {
+		return m.groupsClient.Delete(context.TODO(), group, metav1.DeleteOptions{})
+	}
+
+	// find the user and remove it from the slice
+	userIdx := -1
+	for i, groupUser := range updatedGroup.Users {
+		if groupUser == username {
+			userIdx = i
+			break
+		}
+	}
+
+	var newUsers []string
+	switch userIdx {
+	case -1:
+		return nil
+	case 0:
+		newUsers = updatedGroup.Users[1:]
+	default:
+		newUsers = append(updatedGroup.Users[0:userIdx], updatedGroup.Users[userIdx+1:]...)
+	}
+
+	updatedGroup.Users = newUsers
+
+	_, err = m.groupsClient.Update(context.TODO(), updatedGroup, metav1.UpdateOptions{})
+	return err
+}
+
+// FIXME: work better with annotations
+// FIXME: add retries
```

Comment:
They methods look quite similar, especially if we add more common logic like retries:
```
withRetries:
- Get data
- Modify data
- Update data
```

WDYT?

Reply:
Btw, couldn't we hand the `updatedGroup` down and save two requests in case there are remove AND addings?

Reply:
Alternative approach - use a cache of all groups and use it to do the first `Get`.

It will allow us to keep a unified approach to both the removal and addition actions and we'll avoid the live `Get`.

Reply:
done in the `use lister to get rid off unnecessary GETs` commit

I agree that the removal/addition methods are quite similar although abstracting may not prove that useful because:
1. errors during gets are handled differently
2. the API actions are way different

So at best we would have a "retry" wrapper that needs to be supplied with this custom logic compared to just adding retries to the specific API requests that are a part of this logic.